### PR TITLE
[feat][param_sync]add CollectiveTaskScheduler and speedup first parameter sync 

### PR DESCRIPTION
**Pull request by Yancey1989**
1. add CollectiveTaskScheduler to avoid hang
2. warmup collective group to speedup first parameter sync

### CODE REVIEW

File: chatlearn/synchronizer/parameter_sync.py [hunk 1]
```diff
@@ -998,14 +1053,16 @@ def sync_allgather_multi_threads(
                     _future.result()
                 except Exception as e:
                     raise RuntimeError(f"Parameter sync thread generated an exception: {e}") # pylint: disable=raise-missing-from
-            concurrent.futures.wait(futures)
+                concurrent.futures.wait(futures)
```
Comment (adoda) [inline chatlearn/synchronizer/parameter_sync.py:1056]: 这里把一起等变成串行等？

Comment (Yancey1989) [reply chatlearn/synchronizer/parameter_sync.py]: 我 revert 掉这一行了，应该是 rebase 的时候搞错了。

File: chatlearn/synchronizer/parameter_sync.py [hunk 2]
```diff
@@ -649,15 +691,18 @@ def sync_broadcast_two_stage(self, actors, group_name, requires_grad=None, stage
...
                 sync params of relative rank to other slices of inference model. while {len(actors)}"
+        start_remote_call = time.time()
         for rank, actor in enumerate(actors):
             sync_buffer_rank = self.actor2rank[actors[1]] if rank == 0 and stage2 else 0
             ref = actor.broadcast_parameter_two_stage.remote(
                 self.actor2rank[actor], sync_buffer_rank, rank, send_rank, group_name, pipe_stage, stage2)
             refs.append(ref)
         rets = future.wait(refs, return_output=True)
+        logger.info(f"DEBUG ParameterSync finish broadcast_parameter_two_stage.remote {group_name} using {time.time()-start_remote_call} seconds")
+        logger.info(f"DEBUG ParameterSync {stage_str} finish sync_broadcast_two_stage {group_name} using {time.time()-start_time} seconds")
```
Comment (haolin-nju) [inline chatlearn/synchronizer/parameter_sync.py:702]: `logger.info` mismatches `DEBUG`

File: chatlearn/synchronizer/parameter_sync.py [hunk 3]
```diff
@@ -685,7 +730,7 @@ def sync_alltoall(self, actors, requires_grad=None, filter_fn=None):
             self.set_sync_param_names(actor, actor, requires_grad, filter_fn, param_group="routed", should_map_name=False)
         pipe_stage = self.get_actor_pipe_rank(actors[0])
         refs = []
-        logger.info(f"apply alltoall among {[self.actor2rank[actor] for actor in actors]}")
+        logger.info(f"paply alltoall among {[self.actor2rank[actor] for actor in actors]}")
```
Comment (haolin-nju) [inline chatlearn/synchronizer/parameter_sync.py:733]: typo? paply -> apply

File: chatlearn/synchronizer/parameter_sync.py [hunk 4]
```diff
@@ -738,6 +783,7 @@ def inner_func():
 
     def get_actor_dp_rank(self, actor):
         def inner_func():
+            logger.info(f"DEBUG get_actor_dp_rank: {actor}")
```
Comment (haolin-nju) [inline chatlearn/synchronizer/parameter_sync.py:786]: `logger.info` mismatches `DEBUG`

File: chatlearn/synchronizer/parameter_sync.py [hunk 5]
```diff
@@ -1172,11 +1241,16 @@ def _multi_thread_sync_for_tp_num_mapping_eq_1(
                     logger.info(f"Sending from {[self.actor2rank[send_actor]]} to {[self.actor2rank[actor] for actor in recv_actors]}.")
                     if self._comm_type == PARAM_SYNC_COMM_TYPE.BROADCAST:
                         actor_groups, finalized_group_name = self.create_broadcast_group(send_actor, recv_actors, param_group=param_group)
+                        if dryrun:
+                            continue
                         futures.append(executor.submit(
-                            self.sync_broadcast, actor_groups, finalized_group_name, requires_grad, filter_fn=filter_fn, param_group=param_group
+                            self.sync_broadcast, actor_groups, finalized_group_name, requires_grad, filter_fn=filter_fn, param_group=param_group,
+                            dryrun=dryrun
```
Comment (haolin-nju) [inline chatlearn/synchronizer/parameter_sync.py:1248]: dry_run is always False in this case. Is it the expected behavior?

File: tests/synchronizer/test_scheduler.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        paralel_tasks = next(generator)
+        self.assertEqual([1, 3], paralel_tasks[0].actors)
+        paralel_tasks = next(generator)
+        self.assertEqual([2, 1], paralel_tasks[0].actors)
+        
+        def task_func(task):
+            if task.actors[0] == 2:
+                self.assertEqual("group2", task.group)
+
+        parallel_execute_collective_tasks(tasks, task_func)
```
Comment (haolin-nju) [inline tests/synchronizer/test_scheduler.py:24]: there should be a new line at the end of the file

File: chatlearn/synchronizer/scheduler.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    todo_queue = Queue()
+    pending_queue = []
+    _ = [todo_queue.put(task) for task in tasks]
+
+    while not todo_queue.empty():
+        send_actors_set = set()
+        recv_actors_set = set()
+        list_count = todo_queue.qsize()
+        # re-put it if confilict, otherwise put it to PendingQueue
```
Comment (haolin-nju) [inline chatlearn/synchronizer/scheduler.py:45]: typo? confilict -> conflict

File: tests/synchronizer/test_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+from chatlearn.synchronizer.scheduler import CollectiveTask, collective_task_scheduler, parallel_execute_collective_tasks
+from unittest import TestCase
+class TestCollectiveTaskScheduler(TestCase):
```
Comment (haolin-nju) [inline tests/synchronizer/test_scheduler.py:3]: missing two new lines between import and class definition

**Review by haolin-nju — COMMENTED** [review]
See comments. A common issue is `logger.info` mismatches `DEBUG`.

File: chatlearn/models/base_module.py [hunk 9]
```diff
@@ -495,6 +497,14 @@ def setup_collective_group(self, rank, world_size, backend, group_name):
         self._world_size = world_size
         col.init_collective_group(
             world_size, rank, backend=backend, group_name=group_name)
+    def broadcast_dummy_tensor_send(self, src_rank, group_name):
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply chatlearn/models/base_module.py:500]: done

File: chatlearn/synchronizer/parameter_sync.py (hunk 5 above)
Comment (Yancey1989) [reply chatlearn/synchronizer/parameter_sync.py]: remove the dryrun argument, do not call this function if `dryrun=False`

File: chatlearn/runtime/engine.py [hunk 10]
```diff
@@ -317,6 +317,12 @@ def learn(self):
             logger.info("dump parameters before syncnizing...")
             self.dump_parameters(os.path.join(dump_root_path, "before_sync_parameter"))
         self.timers("sync_parameters").start()
+        if os.getenv("ENABLE_PARAM_SYNC_WARMUP", "ON") == "ON":
```
Comment (adoda) [inline chatlearn/runtime/engine.py:320]: 开关可以统一成Ture或False

File: chatlearn/models/base_module.py [hunk 11]
```diff
@@ -495,6 +497,15 @@ def setup_collective_group(self, rank, world_size, backend, group_name):
...
```
Comment (adoda) [inline chatlearn/models/base_module.py:500]: func间可以加一个换行

File: chatlearn/synchronizer/parameter_sync.py [hunk 12]
```diff
@@ -1001,11 +1049,13 @@ def sync_allgather_multi_threads(
             concurrent.futures.wait(futures)
 
     def sync_alltoall_multi_threads(
-        self, send_actors, max_workers=1, requires_grad=None, filter_fn=None
+        self, send_actors, max_workers=1, requires_grad=None, filter_fn=None, dryrun=False
```
Comment (adoda) [inline chatlearn/synchronizer/parameter_sync.py:1052]: all2all的可以先不修改

**Review by adoda — APPROVED** [review]
LGTM
